pages/instructorEffectiveUser/instructorEffectiveUser.sql

```diff
@@ -13,3 +13,14 @@ SELECT
         WHERE
             e.role <= $authn_role
     ) AS available_uids;
+
+-- BLOCK add_and_enroll
+WITH rows as (
+    INSERT INTO users (uid, name) VALUES ($uid, $uid)
+    ON CONFLICT (uid) DO UPDATE SET uid=$uid /* needed for the return */
+    RETURNING user_id
+)
+INSERT INTO enrollments (user_id, course_instance_id, role)
+    SELECT user_id, $course_instance_id, 'Student'
+    FROM rows
+ON CONFLICT DO NOTHING;
```

Comment:
This code adds a user with the given UID if that user does not exist, and enrolls this new user in the given course instance.

A corner case is when this user already exists, but is not enrolled in the given course instance. In this case, you still get "access denied" (it triggers the `ON CONFLICT`). It seems like the right behavior would be:

- If the user does not exist, create one, and enroll in the course instance as student.
- If the user does exist and is not enrolled in the course instance, enroll in the course instance as student.
- If the user does exist and is enrolled in the course instance, do nothing.

Does that make sense?